backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/FilterPrivilegedDatasetsHandler.java

```diff
@@ -0,0 +1,151 @@
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              // Validate linked dataset in dataset version exists in database records if not then
+              // we will throw the error
+              Set<String> datasetIds = new HashSet<>();
+              for (Artifact dataset : datasets) {
+                String datasetVersionId = dataset.getLinkedArtifactId();
+                if (datasetVersionDatasetMap.containsKey(datasetVersionId)
+                    && datasetVersionDatasetMap.get(datasetVersionId) != null) {
+                  datasetIds.add(datasetVersionDatasetMap.get(datasetVersionId));
+                } else {
+                  return InternalFuture.failedStage(
+                      new InvalidArgumentException(
+                          "Dataset not found for dataset version: " + datasetVersionId));
+                }
+              }
+
+              // If all dataset with linked dataset versions found in database then checking
+              // authorization for the current user for those datasets
+              List<InternalFuture<String>> internalFutures = new LinkedList<>();
+              for (String datasetId : datasetIds) {
+                try {
+                  internalFutures.add(
+                      InternalFuture.completedInternalFuture(datasetId)
+                          .thenCompose(
+                              id ->
+                                  permissionCheck
+                                      .checkPermission(
+                                          Collections.singletonList(datasetId),
+                                          ModelDBActionEnum.ModelDBServiceActions.READ,
+                                          ModelDBResourceEnum.ModelDBServiceResourceTypes.DATASET)
+                                      .thenCompose(
+                                          unused -> InternalFuture.completedInternalFuture(id),
+                                          executor),
+                              executor));
+
+                } catch (PermissionDeniedException ex) {
+                  LOGGER.trace(ex.getMessage());
+                  // We are throws error while creating entities
+                  // if we are fetching entities then we are just skip those datasets from response
+                  // and skipped PermissionDeniedException on fetch
+                  if (errorOut) {
+                    return InternalFuture.failedStage(ex);
+                  }
+                }
+              }
+
+              // If all mapped datasets are allowed for the user then we will filter datasets based
+              // on linked dataset version and allowed dataset ids.
+              InternalFuture<Set<String>> accessibleDatasetIdsFutures =
+                  InternalFuture.sequence(internalFutures, executor)
+                      .thenCompose(
+                          strings -> InternalFuture.completedInternalFuture(new HashSet<>(strings)),
```

Comment:
`strings` is a horrible variable name bc it has nothing to do with what they contain, just their type (which I can already know by checking the type that the IDE already gives me

Reply:
Renamed it with `accessibleDatasetIds`